Reduction getting stuck on large ranges

**Bug summary**

I just pulled latest master, and attempted to run a code including SYCL (sum) reduction. I noticed that the computation was taking unexpectedly long (and I never waited for it to finish). Instead, I created a small repro and noticed that after some threshold size, even an empty kernel "carrying" a reduction objection would take forever to execute.

**To Reproduce**

A minimal repro:

```cpp
#include <sycl/sycl.hpp>

int main(int argc, char **argv) {

  sycl::queue queue;

  unsigned long long res = 0;
  sycl::buffer<unsigned long long> res_b(&res, 1);

  long long len = (argc > 1) ? atoll(argv[1]) : 1 << 24;

  queue.submit([&](sycl::handler &cgh) {
    auto sum_reduction = sycl::reduction(res_b, cgh, sycl::plus<>());
    cgh.parallel_for(sycl::range<1>(len), sum_reduction,
                     [=](sycl::id<1> item, auto &sum) {});
  });
  queue.wait();

  return 0;
}
```

I didn't pass any extra flags to the ACPP build. Then, I created my executable as:
acpp --acpp-targets="generic" -O3  reduce.cpp -o red_acpp

And simply run:
for p in {1..5}; do time ./red_acpp $((1<<(14 + $p))); done

The first 3 runs are fast as expected, but then I am stuck, so Ctrl+C:
```
real    0m0,397s
user    0m0,246s
sys     0m0,151s

real    0m0,348s
user    0m0,220s
sys     0m0,135s

real    0m0,318s
user    0m0,229s
sys     0m0,092s
^C
real    12m59,571s
user    0m0,422s
sys     0m0,124s
```

**Additional diagnostic information**

Few trivial observations:

* I noticed that ACPP recognizes my CPU both as OpenCL as well as OpenMP backend. Same things happens on both backends;
* Substituting sycl::plus with std::plus is of no help;
* Switching to nd_range is of no help;
* Buffer pointing to an array of 1 element instead of a reference is of no help.
* It suffices to get rid of the reduction objects in the cgh.parallel_for to get the expected behaviour, i.e. this doesn't get stuck:
```
  queue.submit([&](sycl::handler &cgh) {
    auto sum_reduction = sycl::reduction(res_b, cgh, sycl::plus<>());
    cgh.parallel_for(sycl::range<1>(len), [=](sycl::id<1> item) {}); });
```

Thanks a lot for the detailed answer @illuhad. Before closing this issue, I would like to point out that when using the latest release 24.02.0, **both the buffer/accessor as well as** the USM version snippet would fail to compile as:
```
error: too few template arguments for class template 'plus'
  q.parallel_for(len, sycl::reduction(res_buff, sycl::plus<>()),
                                                      ^
[]/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/libkernel/functional.hpp:37:30: note: template is declared here
template <typename T> struct plus {
~~~~~~~~~~~~~~~~~~~~~        ^
1 error generated.
```

Okay, so I tried `sycl::plus<unsigned long long>` which would fail to compile with a very long error message starting with:
```
In file included from reduce.cpp:1:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/sycl.hpp:31:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/sycl.hpp:93:
In file included from[]/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/queue.hpp:51:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/handler.hpp:67:
In file included from []/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/hipSYCL/glue/kernel_launcher_factory.hpp:57:
[]/AdaptiveCpp-24.02.0/build/install/bin/../include/AdaptiveCpp/hipSYCL/glue/llvm-sscp/sscp_kernel_launcher.hpp:163:7: error: no matching function for call to object of type 'const (lambda at red_acpp.cpp:15:18)'
      _k(this_item);
      ^~
```

On the other hand, in the develop branch, both `sycl::plus<>` as well as `sycl::plus<unsigned long long>` work as expected.

--- a/include/hipSYCL/runtime/operations.hpp
+++ b/include/hipSYCL/runtime/operations.hpp
@@ -101,6 +101,9 @@ public:
     return make_success();
   }
 
+  virtual std::unique_ptr<requirement>
+  clone_requirement(bool bind_to_same_id) const = 0;
+
   virtual ~requirement(){}
 };
 
@@ -278,6 +281,17 @@ public:
   }
   
   void dump(std::ostream & ostr, int indentation=0) const override;
+
+  virtual std::unique_ptr<requirement>
+  clone_requirement(bool bind_to_same_id) const final override {
+    auto new_req = std::make_unique<buffer_memory_requirement>(
+        _mem_region, _offset, _range, _mode, _target);
+    
+    if(bind_to_same_id)
+      new_req->bind(_bound_embedded_ptr_id);
+
+    return new_req;
+  }
 
 private:
   bool page_ranges_intersect(buffer_data_region::page_range other) const{

--- a/include/hipSYCL/sycl/handler.hpp
+++ b/include/hipSYCL/sycl/handler.hpp
@@ -808,8 +808,16 @@ private:
       // buffers for buffer-accessor reductions
       for(const rt::dag_node_ptr& req : _requirements.get()) {
         auto* op = req->get_operation();
-        if(op->is_requirement())
-          req_list.add_node_requirement(req);
+        if(op->is_requirement()) {
+          auto cloned_op =
+              static_cast<rt::requirement *>(op)->clone_requirement(true);
+
+          req_list.add_requirement(std::move(cloned_op));
+        } else {
+          // Other dependencies that are not requirements should be
+          // covered by the dependency to the previous node that we add
+          // before this for loop.
+        }
       }
       
       previous_event =

